MetaMask/metamask-extension
# Fix migration 88 to handle the case where chainId keys can be undefined

## Explanation

Fixes https://github.com/MetaMask/metamask-extension/issues/20297

At a certain point in history, it was possible to add networks without giving them a chainId. Then, in v10.34.0, we shipped an update to the network controller that would throw an error if a `networkConfiguration` did not have a chain id. This could cause metamask to crash on startup and land users on the error screen.

A fix to this problem is added in this PR in migration 91. It deletes network configurations if they do not have a chain id. There is no risk of hurting user experience here, as these networks have not been usable for a long time.

That fix on its own is insufficient however, because the lack of a chainId in network properties also could have led to a situation where migration 88 errors due to its attempt to validate whether each chainId property key in the various token controllers is hex. This validation ends up passing these keys to bn.js. If a user had once had a network without a chainId, it is possible that they have a string `'undefined'` key in some of the token controllers (where a decimal or hex chainId would have been expected). Passing this key to bn.js results in a thrown error, which results in migrations >= 88 not to run.

So to fix the originally described problem, in addition to migration 91, we need a fix to migration 88. This PR does this by adding logic to migration 88 which will just delete any entries in the tokens controllers that are keyed by `'undefined'` or `undefined`.


## Screenshots/Screencaps

## Manual Testing Steps

1. Install v8.0.0
2. Onboard with the new seed phrase option, but select "Remind me later"
3. Manually add a custom network, but only add the network name and rpc url. Do not add the chainId
4. Update to v10.34.0. MetaMask should not load and you should see the error screen
5. `yarn dist` on this branch. MetaMask should load successfull

## Pre-merge author checklist


## Pre-merge reviewer checklist


If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.
